### ISSUE TITLE
Feature request : enable cv.imencode and cv.imdecode in opencv.js

### ISSUE THREAD

**Issue by pfeatherstone** [opencv/opencv#17535]
Please enable cv.imencode and cv.imdecode in opencv.js while supporting PNG, JPEG and WEBP.

**Comment by ganesh-k13** [comment]
@alalek / @dkurt , the change would be to add `imgcodecs = {'': ['imdecode', 'imencode']}` in `platforms/js/opencv_js.config.py` and update `white_list` below? Or do we do something like `imread` in `modules/js/src/helpers.js`?

**Comment by mshabunin** [comment]
Correct me if I'm wrong, but I thought that OpenCV for JS is built without libpng/libjpeg/... support. Why not use built-in API for this? E.g. [Canvas API for encoding](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toDataURL).

**Comment by pfeatherstone** [comment]
You can indeed use canvas but that seems silly when encoding/decoding are simple vec2vec functions. It seems overkill to create a Dom element to do that. There are libraries like jpegasm which compile C libraries to asm using emscripten. So there is a motivation to do the same in opencv.js.

**Comment by ganesh-k13** [comment]
@mshabunin / @pfeatherstone wont it be better if we use the native code and use a wrapper to call it like [this](https://github.com/opencv/opencv/blob/e2a5a6a05c7ce64911e1e898e986abe8dd26cab6/modules/js/src/core_bindings.cpp#L674)

**Comment by pfeatherstone** [comment]
I guess to do this, you would have to build libpng and libjpeg using emscripten, or whatever opencv.js is built with. There are loads of libraries that do this. Some of the PNG and JPEG libraries for javascript/node use asm instead of pure typescript. We can knock this idea on the head and say it's too faffy in which case we can close the issue.

**Comment by ganesh-k13** [comment]
Oh I see. Or we could do a version of @mshabunin and yours but use some other library for libpng and maintain a version of imdecode and encode in core_bindings.cpp

**Comment by ganesh-k13** [comment]
Ok so people are asking for this: [Answers](https://answers.opencv.org/question/227163/imencode-in-opencvjs/), [SO](https://stackoverflow.com/questions/61751810/opencv-js-mat-imencode) and even a [workaround](https://stackoverflow.com/questions/61477860/cvimdecode-an-image-from-js-to-c-opencv-emscripten).

Couple of things I found out there:
- [CamanJS](http://camanjs.com/) or [JIMP](https://github.com/oliver-moran/jimp), we can do something like what they are doing.
- We can use some built in API for this like shown above.
- Else adapt our own version of same in core_bindings.cpp.

Can we have a discussion on how to implement this? @mshabunin @alalek @berak @pfeatherstone .

**Comment by pfeatherstone** [comment]
I would do something like this: https://github.com/gchudnov/jpeg-asm. It requires a small C, C++ wrapper then a JS binding.

**Comment by ganesh-k13** [comment]
Yeah both CamanJS and JIMP are using jpeg-asm.

**Comment by pfeatherstone** [comment]
Oh sorry. I should have looked closer.

**Comment by mshabunin** [comment]
As I understand there are two problems to be solved: 3rdparty image codec libraries are not compiled for JS (libjpeg, libpng) and `vector<>` type is not supported by current wrapper mechanism.

**Comment by pfeatherstone** [comment]
With regards to compiling for JS, the way to do it is to write two small wrapper functions in C (```encode```, ```decode```), then write a small JS wrapper with the emscripten boiler plate code, as is done in jpeg-asm. So this is already done for jpeg and with a bit of googling i'm sure someone has done it for png, otherwise it shouldn't be hard. 
If vector<> is not allowed, then we have to be grown ups and use C.
So generating ASM code for jpeg, png, webp, tiff and others shouldn't be hard. The main challenge I think is linking with opencv.js. I'm not familiar with the opencv.js codebase to say for sure what the challenges will be then.

**Comment by pfeatherstone** [comment]
We might find it's easier to just use JIMP and not bother putting it in opencv.js. Essentially it becomes syntactic sugar using cv.imencode and cv.imdecode.

**Comment by ganesh-k13** [comment]
I agree, can we use JIMP? From a user perspective they'll have a somewhat native feel by doing `cv.imdecode`.

**Comment by pfeatherstone** [comment]
Need to check that JIMP actually has easy functions like ```encode``` and ```decode``` for multiple codecs, preferably with a synchronous API like opencv.js. If not, then this issue is still relevant.

**Comment by NikkiPriya** [comment]
Hi, I would like to work on this issue, can you help me ways to get started?

**Comment by imrahul361** [comment]
is this issue available? Can i work on this? @pfeatherstone

**Comment by ganesh-k13** [comment]
@alalek / @dkurt , I think it's safe to remove the `good first issue` tag, as I feel porting libpng, etc to here(or calling wrappers) won't be simple.

**Comment by pfeatherstone** [comment]
> is this issue available? Can i work on this? @pfeatherstone

You can do whatever you want mate. I'm not a maintainer. There's no reason why you can submit a PR.

**Comment by imrahul361** [comment]
I am new. Just anyone brief me how to reproduce the issue and other things?

**Comment by braindotai** [comment]
Aren't cv.imencode/imdecode already there in opencvjs? [Here](https://github.com/ucisysarch/opencvjs/blob/7fe821305dffdc3ca6c7edfaa779d045f0406961/test/test_io.js#L70) I can see them being used 🤔

**Comment by DanielBailey-web** [comment]
@braindotai I don't believe that is an official build. It would be nice to have this feature as it gives much more file formats than the native html canvas api does.

**Comment by sanchezcarlosjr** [comment]
In a web worker environment, the Image doesn't work, so the docs don't apply to it. You can use base64, but it is not the best approach because it increases the message size.
Send and receive image bytes would be better than base64.
Suppose the following scenario where node A with Python publishes a message to some web worker.
```python
 while cap.isOpened():
      success, frame = cap.read()
      frame = cv2.cvtColor(cv2.flip(frame, 1), cv2.COLOR_BGR2RGB)
      frame = cv2.resize(frame, (resize_width, resize_height))
      retval, buffer = cv2.imencode('.jpg', frame)
      await client.publish("camera/image", payload=buffer.tobytes(), qos=0)
```
This is better than base64 or using Canvas:
```
receives().pipe(
  map(image => cv.imdecode(image))
)
```

**Comment by lizhen12345** [comment]
i find the [opencv.js](https://docs.opencv.org/4.7.0/d5/d10/tutorial_js_root.html) package is not hvae 'imencode', look this [demo,,,](https://codesandbox.io/s/techstark-opencv-js-demo-page-forked-csvhn8?file=/src/TestPage.jsx) so i want package opencv to opencv.js by myself,,,, so what should i modify。。

**Comment by berak** [comment]
@lizhen12345 , to achieve that, you'd have to wrap & build the whole imgcodecs module (including a good dozen 3rd party libs like libTiff ...) 

as it is now, this does not seem a good idea (far too difficult, will add several mb to the js, etc !)

(my 2 ct:) rather don't, and try to use some 3rd party *js* lib, like the mentioned JIMP package
